### DISCUSSION TITLE
Create posts with chapters

### DISCUSSION THREAD

**Discussion by cat-a-flame** [11ty/eleventy#2243] [Q&A]
Hi there!

I want to create posts with chapters, but not all of them has a chapter. For example:

- Post 1:
• chapter 1
• chapter 2
- Post 2
- Post 3
- Post 4:
• chapter 1
• chapter 2
• chapter 3

At the bottom of these posts, I want to have a pagination with `go to next chapter 'Title of the chapter'` and `go back to previous chapter 'Title of the chapter'`.
Also, I'd like to add a table of content on the very beginning of the post for easier navigation.

My folder structure looks like this:
- Folder for Post 1 containing the chapter.md files
- post2.md
- post3.md
- Folder for Post 4 containing the chapter.md files

I managed to list the related chapters within a folder, but the pagination is only visible on the first page, not on the rest. I also want to make it generic, now it's tied to a specific collection.

I can't really seem to find a solution for this. All of the examples I found are only navigation among posts only.

Edit: updated with more specific details.

**Comment by pdehaan** [comment]
@Stefanye Do you have a public repo or something I can look at?
It might help to see some code.

**Reply by cat-a-flame** [reply]
I managed to make the pagination appear, but only on the first page of the post, not on the rest. 
I've pushed what I have so far so you can check what am I missing, [here is the link](https://github.com/Stefanye/schatten.hu/tree/post-pagination).
The post I want to have chapters is posts/bouquet-preview-app-case-study
The introduction is my first, opening page.

**Reply by pdehaan** [reply]
Sorry, been a bit busy today. I'm seeing an error while building that branch locallly, but I'll keep poking at it.

> > schatten.hu@2.0.0 build
> > eleventy
> 
> [11ty] Problem writing Eleventy templates: (more in DEBUG output)
> [11ty] > Dependency Cycle Found: ___TAG___all -> ./posts/bouquet-preview-app-case-study/introduction.md -> ___TAG___floral -> ./posts/bouquet-preview-app-case-study/introduction.md
> 
> `Error` was thrown:
> [11ty]     Error: Dependency Cycle Found: ___TAG___all -> ./posts/bouquet-preview-app-case-study/introduction.md -> ___TAG___floral -> ./posts/bouquet-preview-app-case-study/introduction.md
>         at /private/tmp/schatten.hu/node_modules/dependency-graph/lib/dep_graph.js:43:17
>         at /private/tmp/schatten.hu/node_modules/dependency-graph/lib/dep_graph.js:292:11
>         at Array.forEach (<anonymous>)
>         at DepGraph.overallOrder (/private/tmp/schatten.hu/node_modules/dependency-graph/lib/dep_graph.js:291:14)
>         at TemplateMap.getMappedDependencies (/private/tmp/schatten.hu/node_modules/@11ty/eleventy/src/TemplateMap.js:143:18)
>         at TemplateMap.cache (/private/tmp/schatten.hu/node_modules/@11ty/eleventy/src/TemplateMap.js:332:30)
>         at TemplateWriter._createTemplateMap (/private/tmp/schatten.hu/node_modules/@11ty/eleventy/src/TemplateWriter.js:242:28)
>         at async TemplateWriter.generateTemplates (/private/tmp/schatten.hu/node_modules/@11ty/eleventy/src/TemplateWriter.js:275:5)
>         at async TemplateWriter.write (/private/tmp/schatten.hu/node_modules/@11ty/eleventy/src/TemplateWriter.js:321:23)
>         at async Eleventy.executeBuild (/private/tmp/schatten.hu/node_modules/@11ty/eleventy/src/Eleventy.js:986:13)
> [11ty] Wrote 0 files in 0.04 seconds (v1.0.0)

**Reply by pdehaan** [reply]
OK, think I almost understand that error message. Looks like a circular loop  w/ collections.
I got the site working by creating a posts/bouquet-preview-app-case-study/bouquet-preview-app-case-study.11tydata.json directory data file with the following:

```json
{
  "eleventyExcludeFromCollections": true
}
```

This will exclude the following files from the "posts" and "floral" collections (and therefore the front page):
- posts/bouquet-preview-app-case-study/closing.md
- posts/bouquet-preview-app-case-study/data.md
- posts/bouquet-preview-app-case-study/introduction.md
- posts/bouquet-preview-app-case-study/used-methods.md

But there is currently no links to the case studies, so I had to manually navigate to http://localhost:8080/posts/bouquet-preview-app-case-study/introduction/ (and there was no [/posts/bouquet-preview-app-case-study/](http://localhost:8080/posts/bouquet-preview-app-case-study/) page).

**Reply by cat-a-flame** [reply]
I have no issues running this branch locally :o I'll check again! 
The post is missing probably cause of the error.

**Reply by cat-a-flame** [reply]
Updated my issue to be more specific!

**Comment by genemars** [comment]
If navigating through pages in the same section is what you are trying to achieve, you can create or filter collections based on the folders files are located in.

**somewhere in your `.eleventy.js` file**, where `eleventyConfig` is available put the following code:
```js
  const fs = require('fs');
  const pagesFolder = './source/pages';
  // get list of sub-folders inside the "pages" folder
  const pageSections = fs.readdirSync(pagesFolder, {withFileTypes: true})
      .filter((dirent) => dirent.isDirectory())
      .map((dirent) => dirent.name);

  // add a collection for each folder
  pageSections.forEach((section) => {
    eleventyConfig.addCollection(`posts_${section}`, function(collection) {
      return collection.getFilteredByGlob(path.join(pagesFolder, `${section}/*.md`))
          .slice().sort((a, b) => +a.data.order > +b.data.order ? 1 : -1);
    });
  });
```
This will create a collection `posts_<folder_name>` for each folder/section.

Then you can add an include file for page navigation (previous page, next page in the same folder/section):

**file: `./source/_inc/page_navitation.html`**
```html
{% assign prevPost = collections | getCollection: page | getPreviousCollectionItem: page %}
{% assign nextPost = collections | getCollection: page  | getNextCollectionItem: page %}

<div -self="size-x1" layout="row center-spread" style="margin-top: 96px; margin-bottom: 56px; padding-left: 8px; padding-right: 8px; font-size: 140%">

    {% if prevPost %}
    <a href="{{ prevPost.url |  url }}">
        <label layout="row center-center">
            <i class="material-icons">arrow_left</i>&nbsp;
            <span style="padding-right: 24px">Previous</span>
        </label>
        <small layout="row center-center">{{ prevPost.data.title }}</small>
    </a>
    {% endif %}

    {% if nextPost %}
    <a href="{{ nextPost.url |  url }}">
        <label layout="row center-center">
            <span style="padding-left: 24px">Next</span>
            <i class="material-icons">arrow_right</i>&nbsp;
        </label>
        <small layout="row center-center">{{ nextPost.data.title }}</small>
    </a>
    {% endif %}

</div>
```

also a `getCollection` filter is required, it will auto-detect and return the collection in which the current page is located:

**file: `./source/_filters/getCollection.js`**
```js
module.exports = (collections, page) => {
  let collection = collections[`posts_blog`]; // default collection
  const path = page.url.split('/');
  if (path.length > 2) {
    const section = 'posts_' + path[path.length - 3];
    if (collections[section]) {
      collection = collections[section];
    }
  }
  return collection;
};
```

**Reply by cat-a-flame** [reply]
Updated my post with more specific details. What you wrote here seems a good idea but I can't make it work. I can't include the page_navigation.html because I keep getting errors during the import. no matter how I specify the path of the file.

**Comment by genemars** [comment]
An example of the code in my previous post is here:
https://zuixjs.github.io/zuix-web-starter
and its repo
https://github.com/zuixjs/zuix-web-starter

**Reply by cat-a-flame** [reply]
In this example, you navigate among different posts, while I have to navigate within one post's chapters. Hard part is to include the pagination for those posts which have more pages, and make it generic.

**Reply by cat-a-flame** [reply]
Also, I can't seem to find where do you import the page_navigation file?

**Reply by genemars** [reply]
https://github.com/Stefanye/schatten.hu/pull/13

**Reply by genemars** [reply]
@Stefanye , see my pull request, I also added the table of contents to the main page layout.

**Comment by genemars** [comment]
@Stefanye, as said, it seems related to [this issue](https://github.com/11ty/eleventy/issues/644#issuecomment-1017407548).
After reading that, I succeed in making it work using your original code from `page-pagination` branch.
I just edited the `posts.json` and removed "posts" from `tags`, like this:
```
{
    "layout": "base.liquid",
    "tags": []
}
```
I could also add the `next` and `previous` link by adding at the end of `.md` files the following code:
```
{% assign chapters = collections.floral %}
{% include 'footer.liquid' %}
```
where `footer.liquid` was a file created in the `_includes` folder:
```
{% if chapters %}
{% assign previousPost = chapters | getPreviousCollectionItem: page %}
{% assign nextPost = chapters | getNextCollectionItem: page %}
{% if previousPost %}Previous Blog Post: <a href="{{ previousPost.url }}">{{ previousPost.data.pageTitle }}</a>{% endif %}
{% if nextPost %}Next Blog Post: <a href="{{ nextPost.url }}">{{ nextPost.data.pageTitle }}</a>{% endif %}
{% endif %}
```

**Reply by cat-a-flame** [reply]
Copied these, but whenever I try to include a file somewhere, I get an error. You can [check my commit](https://github.com/Stefanye/schatten.hu/commit/b10562f92924af17519c50791ea147587f6fb080).

```
$ eleventy --serve
Problem writing Eleventy templates: (more in DEBUG output)
> Having trouble rendering liquid (and markdown) template ./posts/bouquet-preview-app-case-study/introduction.md

`TemplateContentRenderError` was thrown
> ENOENT: Failed to lookup 'footer.liquid' in: _includes,./posts/bouquet-preview-app-case-study, file:./posts/bouquet-preview-app-case-study/introduction.md, 
line:5

`RenderError` was thrown
> ENOENT: Failed to lookup 'footer.liquid' in: _includes,./posts/bouquet-preview-app-case-study

`Error` was thrown:
    Error: ENOENT: Failed to lookup 'footer.liquid' in: _includes,./posts/bouquet-preview-app-case-study
Copied 10 files / Wrote 0 files in 0.19 seconds (v0.12.1)
```

**Reply by genemars** [reply]
Ok, I am making a PR to your repo.

**Reply by genemars** [reply]
See [pull request](https://github.com/Stefanye/schatten.hu/pull/16/files)

**Comment by genemars** [comment]
So, thanks to @Stefanye I now have got a better understanding of collection in *11ty*, still not sure if this is the best solution. Maybe some of you can help providing a better solution.
What I did is creating articles *sections* by adding a `section` tag to each section page, so that these can be listed in the home page like this:

**./source/home.liquid**
```liquid
{% assign sections = collections.sections | sort: 'data.order' %}
{%- for section in sections -%}

    <h2>{{ section.data.title }}</h2>

    {% assign posts = collections[section.data.group] | sort: 'data.order' %}
    {%- for post in posts -%}

   <!-- render the post preview here -->

    {%- endfor -%}

{% endfor %}
```

sections are also sorted based on the `order` field in the front matter.  
To complete the required steps for implementing the navigation, a field `group` is also added to the front matter of each section.  
For example consider the following sections:

**./source/pages/docs.liquid**
```liquid
---
layout: my-layout-1
tags: section
group: documentation
---
```

**./source/pages/templates.liquid**
```liquid
---
layout: my-layout-2
tags: section
group: templates
---
```

Section files can eventually also contain the code for listing articles in their target group, as defined in the front matter (and sorted by the `order` field in this example):

```liquid
{% assign posts = collections[group] | sort: 'data.order' %}
{%- for post in posts -%}
<!-- Render post preview here -->
{%- endfor -%}
```

So, articles of each section can be placed anywhere, also organized in sub-folders, and the only thing required to be included in a section is adding both `group` and `tags` fields to the front matter of the article, with the same value of the section we want the article to be listed in:

**./source/pages/documentation/introduction.md**
```md
---
layout: my-doc-layout
tags: documentation examples ...
group: documentation
---
```

So, we can also have page navigation by adding to the common footer of the pages the following code:

**./source/_inc/page_navigation.liquid**
```liquid
{% if group %}

  {% assign prevPost = collections[group] | sort: 'data.order' | getPreviousCollectionItem: page %}
  {% assign nextPost = collections[group] | sort: 'data.order' | getNextCollectionItem: page %}

  {% if prevPost %}
  <a href="{{ prevPost.url |  url }}">Prev: {{ prevPost.data.title }}</a>
  {% endif %}

  {% if nextPost %}
  <a href="{{ nextPost.url |  url }}">Next: {{ nextPost.data.title }}</a>
  {% endif %}

{% endif %}
```

I've also updated the code of the example website I provided in my [previous message](https://github.com/11ty/eleventy/discussions/2243#discussioncomment-2234101) and it's working so far.

@Stefanye let me know if this works for you as well. =)
